upload-notebooks/upload-notebook.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
+
+# Upload notebook to TileDB Cloud
+#
+# Usage:
+#     python upload-notebook.py <path/to/notebook.ipynb>
+#     python upload-notebook.py --help
+#
+# Variables read from the environment:
+#   - TILEDB_CLOUD_TOKEN: API token for TileDB Cloud account (required)
+#   - TILEDB_CLOUD_NAMESPACE: Namespace for TileDB Cloud (required)
+#   - TILEDB_CLOUD_STORAGE_PATH: Storage path on TileDB Cloud (optional)
+#   - TILEDB_CLOUD_STORAGE_CREDENTIAL_NAME: Storage credentials to use on TileDB Cloud (optional)
+
+
+import argparse
+import os
+import sys
+
+import tiledb.cloud
+
+# Process input
+parser = argparse.ArgumentParser(
+    description="Upload notebook to TileDB Cloud",
+    epilog="TileDB Cloud options may also be set as environment variables, but values set via command line flags will take priority",
+)
+parser.add_argument(
+    "notebook",
+    help="Notebook file to upload (extension .ipynb)",
+)
+parser.add_argument(
+    "--tiledb-cloud-token",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_TOKEN"),
+    help="API token for TileDB Cloud account",
+)
+parser.add_argument(
+    "--tiledb-cloud-namespace",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_NAMESPACE"),
+    help="Namespace for TileDB Cloud",
+)
+parser.add_argument(
+    "--tiledb-cloud-storage-path",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_STORAGE_PATH"),
+    help="Storage path on TileDB Cloud",
+)
+parser.add_argument(
+    "--tiledb-cloud-storage-credential-name",
+    required=False,
+    default=os.environ.get("TILEDB_CLOUD_STORAGE_CREDENTIAL_NAME"),
+    help="Storage credentials to use on TileDB Cloud",
+)
+args = parser.parse_args()
+
+notebook = args.notebook
+token = args.tiledb_cloud_token
+namespace = args.tiledb_cloud_namespace
+storage_path = args.tiledb_cloud_storage_path
+storage_credential_name = args.tiledb_cloud_storage_credential_name
+
+# Verify input
+if not os.path.exists(notebook):
+    sys.stderr.write("Error: Notebook file does not exist: %s\n" % (notebook))
+    sys.exit(1)
+
+notebook_name, ext = os.path.splitext(notebook)
+notebook_name = os.path.basename(notebook_name)
+if ext != ".ipynb":
+    sys.stderr.write(
+        "Error: Notebook file extension must be '.ipynb', not '%s'\n" % (ext)
+    )
+    sys.exit(1)
+
+if token is None:
+    sys.stderr.write("Error: Missing API token\n")
+    sys.stderr.write(
+        "You must define the token as the env var TILEDB_CLOUD_TOKEN or pass it to --tiledb-cloud-token\n"
+    )
+    sys.exit(1)
+
+if namespace is None:
+    sys.stderr.write("Error: Missing namespace\n")
+    sys.stderr.write(
+        "You must define the namespace as the env var TILEDB_CLOUD_NAMESPACE or pass it to --tiledb-cloud-namespace\n"
+    )
+    sys.exit(1)
+
+# Login and upload
+tiledb.cloud.login(token=token)
+user = tiledb.cloud.user_profile()
+sys.stderr.write("Info: Logged into TileDB Cloud as %s\n" % (user.username))
```

Comment:
I noticed is that calling `user_profile()` requires a `user:read`-scoped token, whereas creating an array requires the `array:admin` scope. Since it's not currently possible to create a token with multiple scopes, the action as implemented requires a `*`-scoped token, which isn't ideal. 

Since you're only using `user_profile()` to log the username, I think it would be better to remove that so a more limited token can be used. 

 I [verified this works](https://github.com/TileDB-Inc/TileDB-Cloud-SOMA/actions/runs/4725911543/jobs/8401737212) using an `array:admin` token with a [fork of this action](https://github.com/aaronwolen/tiledb-github-actions/commit/73ea9a66d0d0d0757c3ab698b40a4b85f01cefd8) that drops the `user_profile()` call.

 CC'ing @snagles (who helped me test this out) and @shelnutt2 to get their thoughts on this.

Reply:
@aaronwolen I think this is a great suggestion! I always prefer to limit the scope of API tokens when the minimal required scope can be determined. I also tested and confirmed that a TileDB token with scope `array:admin` is sufficient to upload a notebook, so I've removed the call to `user_profile()`